Video Streaming Menu (#123)
smartdevicelink/sdl_app_library_guides: docs/Video Streaming for Navigation Apps/Menus/index.md

# Video Streaming Menu

> 

Reviewer: There needs to be an introduction here explaining the general concept of SDL menus. I'm thinking of something like "When building your video-streaming navigation app, you can either create a menu using your own video streaming UI or use the built-in SDL menu system. The SDL menu system allows you to display a menu structure for users to select menu options or submenus. For more information about the SDL menu system, see [link here]. It's recommended to use the built-in SDL menu system in order to have better performance, automatic driver distraction support – such as list limitations and text sizing, and more.

To open the SDL built-in menu from your video streaming UI, see 'Opening the Built-In Menu' below."

Author: added

Author: added

> ## Open Built-In Menu

Reviewer: ```suggestion
## Opening the Built-In Menu
```

> The Show Menu RPC allows you to open the menu programmatically. When creating a projection applicaiton it is recommended to still use the built-in menu rather then creating a custom one. Since there is no guarantee the built-in menu button will be visble in a projection application we recommned creating a custom button and call the `ShowMenu` RPC.

Reviewer: ```suggestion
The Show Menu RPC allows you to open the menu programmatically. That way, you can open the menu from your own UI.
```

### Show Top Level Menu

> To show the top level menu use the `screenManger`s `openMenu` function.

Reviewer: ```suggestion
To show the top level menu use `sdlManager.screenManager.openMenu`.
```

@![iOS]
##### Objective-C
```objc

> [self.sdlManger.screenManager openMenu];

Reviewer: ```suggestion
[self.sdlManager.screenManager openMenu];
```

```

> ##### Swift

Reviewer: ```suggestion

##### Swift
```

```swift
self.sdlManager.screenManager.openMenu()
```
!@

@![android, javaSE, javaEE]
```java

> toDO - add example 

Reviewer: ```suggestion
// TODO - add example 
```

```
!@

> ### Show Sub Menu

Reviewer: ```suggestion
### Show Sub-Menu
```

> Showing a sub menu is also possible with the new Show App Menu RPC. To open a certain sub menu simply pass the parent cell that contains sub cells. If the cell has no sub cells the RPC will fail. 

Reviewer: ```suggestion
You can also open the menu directly to a sub-menu. This is further down the tree than the top-level menu. To open a sub-menu, pass a cell that contains sub-cells. If the cell has no sub-cells the method call will fail. 
```

@![iOS]
##### Objective-C
```objc
[self.sdlManager.screenManager openSubmenu:(<#CellWithSubCells#>)];
```

> ##### Swift

Reviewer: ```suggestion

##### Swift
```

```swift
self.sdlManager.screenManager.openSubmenu(<#CellWithSubCells#>)
```
!@

@![android, javaSE, javaEE]
```java

> toDO - add example 

Reviewer: ```suggestion
// TODO - add example 
```

```
!@

## Close Application

> Video Streaming / Projection application need a way to close the application if not utilizing  `ShowMenu` . If you are utizling the `ShowMenu` RPC it is highly reccomned to not use the `CloseApplication` RPC.  If you chose to create a custom menu the  `CloseApplication`  may be useful for your users.

Reviewer: ```suggestion
If you choose to not use the built-in SDL menu system and instead to use your own menu UI, you need to have a way for users to close your application. This should be done through a menu option in your UI that sends the `CloseApplication` RPC. 

This RPC is unnecessary if you are using `OpenMenu` because OEMs will take care of providing a close button into your menu themselves.
```

@![iOS]
##### Objective-C
```objc

> SDLCloseApplication *closeApp = [[SDLCloseApplication alloc] init];

Reviewer: ```suggestion
SDLCloseApplication *closeRPC = [[SDLCloseApplication alloc] init];
```

[self.sdlManager sendRequest:closeApp];
```

> ##### Swift

Reviewer: ```suggestion

##### Swift
```

```swift

> let closeApp = SDLCloseApplication()

Reviewer: ```suggestion
let closeRPC = SDLCloseApplication()
```

# Video Streaming Menu
When building a video-streaming navigation application, you can choose to create a custom menu using your own UI or use the built-in SDL menu system. The SDL menu allows you to display a menu structure so users can select menu options or submenus. For more information about the SDL menu system, see [menus](https://smartdevicelink.com/en/guides/iOS/displaying-a-user-interface/main-menu/i). It's recommended to use the built-in SDL menu system to have better performance, automatic driver distraction support - such as list limitations and text sizing, and more.

To open the SDL built-in menu from your video streaming UI, see 'Opening the Built-In Menu' below.

## Opening the Built-In Menu
The Show Menu RPC allows you to open the menu programmatically. That way, you can open the menu from your own UI.

### Show Top Level Menu
To show the top level menu use `sdlManager.screenManager.openMenu`.

@![iOS]
##### Objective-C
```objc
[self.sdlManager.screenManager openMenu];
```

##### Swift
```swift
self.sdlManager.screenManager.openMenu()
```
!@

@![android, javaSE, javaEE]
```java
// TODO - add example 
```
!@

### Show Sub-Menu

> You can also open the menu directly to a sub-menu. This is further down the tree than the top-level menu. To open a sub-menu, pass a cell that contains sub-cells. If the cell has no sub-cells the method call will fail. 

Reviewer: Also note that the cell needs to be in the `sdlManager.screenManager.menu` array.

Author: added

@![iOS]
##### Objective-C
```objc
[self.sdlManager.screenManager openSubmenu:(<#CellWithSubCells#>)];
```

##### Swift
```swift
self.sdlManager.screenManager.openSubmenu(<#CellWithSubCells#>)
```
!@

@![android, javaSE, javaEE]
```java
// TODO - add example 
```
!@

## Close Application

> If you choose to not use the built-in SDL menu system and instead to use your own menu UI, you need to have a way for users to close your application. This should be done through a menu option in your UI that sends the `CloseApplication` RPC. 

Reviewer: ```suggestion
If you choose to not use the built-in SDL menu system and instead want to use your own menu UI, you need to have a way for users to close your application. This should be done through a menu option in your UI that sends the `CloseApplication` RPC.
```

> This RPC is unnecessary if you are using `OpenMenu` because OEMs will take care of providing a close button into your menu themselves.

Reviewer: ```suggestion
!!! NOTE
This RPC is unnecessary if you are using `OpenMenu` because OEMs will take care of providing a close button into your menu themselves.
!!!
```

To open the SDL built-in menu from your video streaming UI, see 'Opening the Built-In Menu' below.

## Opening the Built-In Menu
The Show Menu RPC allows you to open the menu programmatically. That way, you can open the menu from your own UI.

### Show Top Level Menu
To show the top level menu use `sdlManager.screenManager.openMenu`.

@![iOS]
##### Objective-C
```objc
[self.sdlManager.screenManager openMenu];
```

##### Swift
```swift
self.sdlManager.screenManager.openMenu()
```
!@

@![android, javaSE, javaEE]
```java
// TODO - add example 
```
!@

### Show Sub-Menu
You can also open the menu directly to a sub-menu. This is further down the tree than the top-level menu. To open a sub-menu, pass a cell that contains sub-cells. If the cell has no sub-cells the method call will fail. 

!!! Note

> The sub-cell you use in `openSubMenu` must be included in  `sdlManager.screenManager.menu` array. Please make sure it is in the array before trying to open the sub-cell. If it is not included in the array the method call will fail.

Reviewer: ```suggestion
The sub-cell you use in `openSubMenu` must be included in `sdlManager.screenManager.menu` array. If it is not included in the array, the method call will fail.
```
